### PR TITLE
Use functions in episode 4

### PR DESCRIPTION
**Pull request by svenvanderburg**

### CODE REVIEW

File: episodes/4-advanced-layer-types.Rmd [hunk 1]
```diff
@@ -411,16 +426,21 @@ outputs = keras.layers.Dense(10)(x)
...
-x = keras.layers.Conv2D(50, (3, 3), activation='relu')(inputs)
-x = keras.layers.MaxPooling2D((2, 2))(x)
-x = keras.layers.Conv2D(50, (3, 3), activation='relu')(x)
-x = keras.layers.MaxPooling2D((2, 2))(x)
-x = keras.layers.Conv2D(50, (3, 3), activation='relu')(x)
-x = keras.layers.Flatten()(x)
-x = keras.layers.Dense(50, activation='relu')(x)
-outputs = keras.layers.Dense(10)(x)
-model = keras.Model(inputs=inputs, outputs=outputs, name="cifar_model")
+def create_nn():
```
Comment (cpranav93) [inline episodes/4-advanced-layer-types.Rmd:429]: Would it be a better idea to give this function a different name? maybe create_nn_with_extra_layer()?

File: episodes/4-advanced-layer-types.Rmd [hunk 2]
```diff
@@ -411,16 +426,21 @@ outputs = keras.layers.Dense(10)(x)
...
+    x = keras.layers.MaxPooling2D((2, 2))(x) #
+    x = keras.layers.Conv2D(50, (3, 3), activation='relu')(x) # estra layer
+    x = keras.layers.Flatten()(x)
+    x = keras.layers.Dense(50, activation='relu')(x) # a new Dense layer
+    outputs = keras.layers.Dense(10)(x)
+
+    model = keras.Model(inputs=inputs, outputs=outputs, name="cifar_model")
+    return model
+
+model = create_nn()
```
Comment (cpranav93) [inline episodes/4-advanced-layer-types.Rmd:443]: Rename here as well if we rename the function

File: episodes/4-advanced-layer-types.Rmd [hunk 3]
```diff
@@ -525,19 +542,21 @@ In practice, however, dropout is computationally a very elegant solution which d
...
-x = keras.layers.MaxPooling2D((2, 2))(x)
-x = keras.layers.Conv2D(50, (3, 3), activation='relu')(x)
-x = keras.layers.Dropout(0.8)(x) # This is new!
-x = keras.layers.Flatten()(x)
-x = keras.layers.Dense(50, activation='relu')(x)
-outputs = keras.layers.Dense(10)(x)
-
-model_dropout = keras.Model(inputs=inputs, outputs=outputs, name="cifar_model")
-
+def create_nn():
```
Comment (cpranav93) [inline episodes/4-advanced-layer-types.Rmd:545]: Same as before, a new name?

File: episodes/4-advanced-layer-types.Rmd [hunk 4]
```diff
@@ -629,28 +643,33 @@ This is where the test loss is lowest.
 - NB2: In the real world you should do this with a validation set and not with the test set!
 
 ```python
+def create_nn(dropout_rate):
```
Comment (cpranav93) [inline episodes/4-advanced-layer-types.Rmd:646]: As above

File: episodes/4-advanced-layer-types.Rmd [hunk 5]
```diff
@@ -629,28 +643,33 @@ This is where the test loss is lowest.
...
-   outputs = keras.layers.Dense(10)(x)
-   model_dropout = keras.Model(inputs=inputs, outputs=outputs, name="cifar_model")
-   model_dropout.compile(optimizer='adam',
-             loss=keras.losses.SparseCategoricalCrossentropy(from_logits=True),
-             metrics=['accuracy'])
-   model_dropout.fit(train_images, train_labels, epochs=20,
-                   validation_data=(test_images, test_labels))
-   test_loss, test_acc = model_dropout.evaluate(test_images,  test_labels)
-   test_losses.append(test_loss)
+    model_dropout = create_nn(dropout_rate)
```
Comment (cpranav93) [inline episodes/4-advanced-layer-types.Rmd:663]: as above

**Review by cpranav93 — COMMENTED** [review]
The use of functions helps in not retyping code especially the compile model function. I am not sure how much the create_nn function helps as we need to redefine it for each instance (dropout, additional layer, etc). Also, it would be good to rename the function for each redefinition to differentiate the different methods.

File: episodes/4-advanced-layer-types.Rmd (hunk 1 above)
Comment (svenvanderburg) [reply episodes/4-advanced-layer-types.Rmd]: Totally agree! Changed it

**Comment by svenvanderburg** [comment]
I agree that the `create_nn` function does not add as much as the `plot_history` and `compile_model` functions. However, I do think that it shows good practice and conceptually puts all network creation code in one function which makes the explanation a bit easier I think. And it is in line with the other episodes.
